app/domain/services/assignment_exercise_requests.rb

```diff
@@ -0,0 +1,271 @@
+module Services::AssignmentExerciseRequests
+  DEFAULT_NUM_PES_PER_BOOK_CONTAINER = 3
+  DEFAULT_NUM_SPES_PER_K_AGO = 1
+
+  K_AGOS_TO_LOAD = [ 1, 2, 3, 4, 5 ]
+  NON_RANDOM_K_AGOS = [ 1, 3, 5 ]
+
+  MIN_SEQUENCE_NUMBER_FOR_RANDOM_AGO = 5
+
+  def get_excluded_exercises_by_student_uuid(assignments, current_time: Time.current)
+    # Get exercise exclusions for each course
+    course_uuids = assignments.map(&:course_uuid)
+    course_exclusions_by_course_uuid = Course.where(uuid: course_uuids).pluck(
+      :uuid,
+      :global_excluded_exercise_uuids,
+      :course_excluded_exercise_uuids,
+      :global_excluded_exercise_group_uuids,
+      :course_excluded_exercise_group_uuids
+    ).index_by(&:first)
+
+    # Get assignments that should have hidden feedback for each student
+    student_uuids = assignments.map(&:student_uuid)
+    no_feedback_yet_assignments = Assignment
+                                    .where(student_uuid: student_uuids)
+                                    .where(Assignment.arel_table[:feedback_at].gt(current_time))
+                                    .pluck(:student_uuid, :assigned_exercise_uuids)
+
+    # Convert excluded exercise uuids to group uuids
+    assigned_exercise_uuids = no_feedback_yet_assignments.flat_map(&:second)
+    assigned_exercise_group_uuid_by_uuid = Exercise.where(uuid: assigned_exercise_uuids)
+                                                   .pluck(:uuid, :group_uuid)
+                                                   .to_h
+
+    # Convert exclusion group uuids to uuids
+    excluded_exercise_group_uuids = course_exclusions_by_course_uuid.values.flat_map(&:fourth) +
+                                    course_exclusions_by_course_uuid.values.flat_map(&:fifth) +
+                                    assigned_exercise_group_uuid_by_uuid.values
+    excluded_exercise_uuids_by_group_uuid = Hash.new { |hash, key| hash[key] = [] }
+    Exercise.where(group_uuid: excluded_exercise_group_uuids)
+            .pluck(:group_uuid, :uuid)
+            .each do |group_uuid, uuid|
+      excluded_exercise_uuids_by_group_uuid[group_uuid] << uuid
+    end
+
```

Comment:
seems like we could pre-calculate and cache this somewhere, might give us a small speedup